libraries/botbuilder-dialogs-adaptive/src/actions/forEach.ts

```diff
@@ -5,165 +5,78 @@
+        dc.state.setValue(INDEX, -1);
+        return await this.nextItem(dc);
     }
 
-    public getDependencies(): Dialog[] {
-        return this.actions;
+    protected async onBreakLoop(dc: DialogContext, actionScopeResult: ActionScopeResult): Promise<DialogTurnResult> {
+        return await dc.endDialog();
     }
 
-    public async beginDialog(sequence: SequenceContext, options: ForEachOptions): Promise<DialogTurnResult> {
-        // Ensure planning context
-        if (!(sequence instanceof SequenceContext)) { throw new Error(`${this.id}: should only be used within an AdaptiveDialog.`) }
-        if (!this.list) { throw new Error(`${this.id}: no list expression specified.`) }
-
-        // Unpack options
-        let { list, offset } = options;
-        if (list == undefined) { list = this.list.evaluate(this.id, sequence.state) }
-        if (offset == undefined) { offset = 0 }
-
-        // Get next page of items
-        const item = this.getItem(list, offset);
-
-        // Update current plan
-        if (item !== undefined) {
-            sequence.state.setValue(this.valueProperty, item);
-            sequence.state.setValue(this.indexProperty, offset);
-            const changes: ActionChangeList = {
-                changeType: ActionChangeType.insertActions,
-                actions: []
-            };
-            this.actions.forEach((action) => changes.actions.push({ dialogStack: [], dialogId: action.id }));
-
-            // Add a call back into forEachPage() at the end of repeated actions.
-            // - A new offset is passed in which causes the next page of results to be returned.
-            changes.actions.push({
-                dialogStack: [],
-                dialogId: this.id,
-                options: {
-                    list: list,
-                    offset: offset + 1
-                }
-            });
-            sequence.queueChanges(changes);
-        }
+    protected async onContinueLoop(dc: DialogContext, actionScopeResult: ActionScopeResult): Promise<DialogTurnResult> {
+        return await this.nextItem(dc);
+    }
 
-        return await sequence.endDialog();
+    protected async onEndOfActions(dc: DialogContext, result?: any): Promise<DialogTurnResult> {
+        return await this.nextItem(dc);
     }
 
-    private getItem(list: any[]|object, index: number): any {
-        if (Array.isArray(list)) {
-            if (index < list.length) {
-                return list[index];
-            }
-        } else if (typeof list === 'object') {
-            let i = 0;
-            for (const key in list) {
-                if (list.hasOwnProperty(key)) {
-                    if (i == index) {
-                        return list[key];
-                    }
-                    i++;
-                }
-            }
+    protected async nextItem(dc: DialogContext): Promise<DialogTurnResult> {
+        const itemsProperty = new ExpressionEngine().parse(this.itemsProperty);
```

Comment:
>ew ExpressionEngine().parse(this.itemsProperty); [](start = 31, length = 48)

Again, seems like we can parse this just once.

Reply:
> > ew ExpressionEngine().parse(this.itemsProperty); [](start = 31, length = 48)
> 
> Again, seems like we can parse this just once.

Cached parsed items property expression.